RecoBTag/DeepFlavour/plugins/DeepFlavourTagInfoProducer.cc

```diff
@@ -0,0 +1,318 @@
+              { return deep::sv_vertex_comparator(sva, svb, pv); });
+    // fill features from secondary vertices
+    for (const auto & sv : svs_sorted) {
+      if (reco::deltaR(sv, jet) > jet_radius_) continue;
+      else {
+        features.sv_features.emplace_back();
+        // in C++17 could just get from emplace_back output
+        auto & sv_features = features.sv_features.back();
+        deep::sv_features_converter(sv, pv, jet, sv_features);
+      }
+    }
+
+    // stuff required for dealing with pf candidates 
+    math::XYZVector jet_dir = jet.momentum().Unit();
+    GlobalVector jet_ref_track_dir(jet.px(),
+                                   jet.py(),
+                                   jet.pz());
+
+    std::vector<sorting::sortingClass<size_t> > c_sorted, n_sorted;
+
+    deep::TrackInfoBuilder trackinfo(track_builder);
+
+    // unsorted reference to sv
+    const auto & svs_unsorted = *svs;     
+    // fill collection, from DeepTNtuples plus some styling
+    for (unsigned int i = 0; i <  jet.numberOfDaughters(); i++){
+        auto cand = dynamic_cast<const reco::Candidate *>(jet.daughter(i));
+        if(cand){
+          if (cand->charge() != 0) {
+            trackinfo.buildTrackInfo(cand,jet_dir,jet_ref_track_dir,pv);
+            c_sorted.emplace_back(i, trackinfo.getTrackSip2dSig(),
+                    -deep::mindrsvpfcand(svs_unsorted,cand), cand->pt()/jet.pt());
+          } else {
+            n_sorted.emplace_back(i, -1,
+                    -deep::mindrsvpfcand(svs_unsorted,cand), cand->pt()/jet.pt());
+          }
+        }
+    }
+
+    // sort collections (open the black-box if you please) 
+    std::sort(c_sorted.begin(),c_sorted.end(),
+      sorting::sortingClass<std::size_t>::compareByABCInv);
+    std::sort(n_sorted.begin(),n_sorted.end(),
+      sorting::sortingClass<std::size_t>::compareByABCInv);
+
+    std::vector<size_t> c_sortedindices,n_sortedindices;
+   
+    // this puts 0 everywhere and the right position in ind 
+    c_sortedindices=sorting::invertSortingVector(c_sorted);
+    n_sortedindices=sorting::invertSortingVector(n_sorted);
+
+    // set right size to vectors
+    features.c_pf_features.clear();
+    features.c_pf_features.resize(c_sorted.size());
+    features.n_pf_features.clear();
+    features.n_pf_features.resize(n_sorted.size());
+
+
+
+  for (unsigned int i = 0; i <  jet.numberOfDaughters(); i++){
+
+    // get pointer and check that is correct
+    auto cand = dynamic_cast<const reco::Candidate *>(jet.daughter(i));
+    if(!cand) continue;
+    auto packed_cand = dynamic_cast<const pat::PackedCandidate *>(cand);
+    auto reco_cand = dynamic_cast<const reco::PFCandidate *>(cand);
+
+    float drminpfcandsv = deep::mindrsvpfcand(svs_unsorted, cand);
+    
+    if (cand->charge() != 0) {
+      // is charged candidate
+      auto entry = c_sortedindices.at(i);
+      // build track info
+      trackinfo.buildTrackInfo(cand,jet_dir,jet_ref_track_dir,pv);
```

Comment:
it looks like a repeated call to buildTrackInfo for the same candidates in just about 50 lines.
This is not a free operation. Perhaps it is better to cache or compute the features upfront